### PR TITLE
c# interface WIP

### PR DESCRIPTION
**Pull request by lefavilla**
C# Interface in respect to the gd class interface

added Todo line 139

@ricardoalcantara @Shin-NiL maybe you need to tweak it a bit to a more commoner version.

### CODE REVIEW

**Comment by ricardoalcantara** [comment]
Could you follow the Godot C# Stylie guideline https://docs.godotengine.org/en/stable/getting_started/scripting/c_sharp/c_sharp_style_guide.html

Also you could clean up a bit your prints like this:

GD.Print("AdMob:" + admobSingleton);
GD.Print("AdMob:" + admobSingleton);
GD.Print("AdMob:" + admobSingleton);

**Comment by Shin-NiL** [comment]
As I said before I'm not a C# guy, but is there any way on the C# script to access the AdMob node from GDScript?
I've created this script on the admob-lib directory just to work as a visual helper on the editor, if it works on the C# side it still works as intended.

**Comment by lefavilla** [comment]
> 
> 
> As I said before I'm not a C# guy, but is there any way on the C# script to access the AdMob node from GDScript?
> I've created this script on the admob-lib directory just to work as a visual helper on the editor, if it works on the C# side it still works as intended.
@Shin-NiL 
I have no Idea. I know that one can use cross scripting methods by using object.call and object.calldeferred. But as I see it now, we c# guys dont use the node system like the gd pops. You could that .cs for the c# guys. Everyone on the c# side knows how to use it in a monoproject.

Maybe i write down a little .md helper text in how to use this.



> 
> 
> Could you follow the Godot C# Stylie guideline https://docs.godotengine.org/en/stable/getting_started/scripting/c_sharp/c_sharp_style_guide.html
> 
> Also you could clean up a bit your prints like this:
> 
> GD.Print("AdMob:" + admobSingleton);
> GD.Print("AdMob:" + admobSingleton);
> GD.Print("AdMob:" + admobSingleton);

@ricardoalcantara 
Sure I will clean up and repush. I refined the classes based on your practices.

File: admob-lib/AdMob.cs [hunk 1]
```diff
@@ -20,9 +24,9 @@ public class AdMob : Node {
 	#region Properties
 	[Export] bool isReal;
 	[Export] bool bannerOnTop = true;
-	[Export] String bannerId;
-	[Export] String interstitialId;
-	[Export] String rewardedId;
+	[Export] String bannerId = "MISSING_ID";
+	[Export] String interstitialId = "MISSING_ID";
```
Comment (lefavilla) [inline admob-lib/AdMob.cs:28]: Her should be the test ad id provided by google ads.

File: admob-lib/AdMob.cs [hunk 2]
```diff
@@ -20,9 +24,9 @@ public class AdMob : Node {
 	#region Properties
 	[Export] bool isReal;
 	[Export] bool bannerOnTop = true;
-	[Export] String bannerId;
-	[Export] String interstitialId;
-	[Export] String rewardedId;
+	[Export] String bannerId = "MISSING_ID";
```
Comment (lefavilla) [inline admob-lib/AdMob.cs:27]: Her should be the test ad id provided by google ads.

**Comment by lefavilla** [comment]
> Could you follow the Godot C# Stylie guideline https://docs.godotengine.org/en/stable/getting_started/scripting/c_sharp/c_sharp_style_guide.html


I overthinked it a bit. I can't follow the guideline completely from godot. That formatting can be done by someone who cares for. I follow the strict guideline from my company. Sorry.

**Comment by ricardoalcantara** [comment]
> I overthinked it a bit. I can't follow the guideline completely from godot. That formatting can be done by someone who cares for. I follow the strict guideline from my company. Sorry.

This is not how you should contribute with the community, here is like another company there is guidelines to follow, if you don't follow them your PR will be rejected.

I made this whole IAP Interface for C# https://github.com/godotengine/godot-demo-projects/tree/master/mono/android_iap/GodotGooglePlayBilling

And they were very rigorous with guideline
https://github.com/godotengine/godot-demo-projects/pull/498
 
Anyway if you just throw your code for someone to clean it up, I prefer to make it myself, it will take me almost the same amount of time and I will not be worried to have to change someone's code.

@Shin-NiL 
You can perfectly work with GDScript and C# together, it's easier creating an interface, as I did for IAP, since this is your repository it's up to you accept that but I would not accept anything that doesn't follow the guideline.

**Comment by lefavilla** [comment]
> 
> 
> @lefavilla
> 
> > I overthinked it a bit. I can't follow the guideline completely from godot. That formatting can be done by someone who cares for. I follow the strict guideline from my company. Sorry.
> 
> This is not how you should contribute with the community, here is like another company there is guidelines to follow, if you don't follow them your PR will be rejected.
> 
> I made this whole IAP Interface for C# https://github.com/godotengine/godot-demo-projects/tree/master/mono/android_iap/GodotGooglePlayBilling
> 
> And they were very rigorous with guideline
> [godotengine/godot-demo-projects#498](https://github.com/godotengine/godot-demo-projects/pull/498)
> 
> Anyway if you just throw your code for someone to clean it up, I prefer to make it myself, it will take me almost the same amount of time and I will not be worried to have to change someone's code.
> 
> @Shin-NiL
> You can perfectly work with GDScript and C# together, it's easier creating an interface, as I did for IAP, since this is your repository it's up to you accept that but I would not accept anything that doesn't follow the guideline.

Ok sign on that. I will edit my prettier script later on to match the guideline. I hoped that you have an ready to go style guide script ready. That would had saved some work for me. But anyway, I will fix this.

**Comment by lefavilla** [comment]
@ricardoalcantara could you maybe review the style? I'm not 100% sure that this is okay like this. Tried to follow the guideline.

File: admob-lib/AdMob.cs [hunk 3]
```diff
@@ -0,0 +1,251 @@
+using Godot;
+using System;
+
+///AUTHOR: Alexander Jungert - 28/08/2020
```
Comment (ricardoalcantara) [inline admob-lib/AdMob.cs:4]: Why did you leave that comment, do you claim ownership, this lib is MIT.

File: admob-lib/AdMob.cs [hunk 4]
```diff
@@ -0,0 +1,251 @@
+using Godot;
+using System;
+
+///AUTHOR: Alexander Jungert - 28/08/2020
+public class AdMob : Node 
+    {
```
Comment (ricardoalcantara) [inline admob-lib/AdMob.cs:6]: There is space before {

File: admob-lib/AdMob.cs [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+///AUTHOR: Alexander Jungert - 28/08/2020
+public class AdMob : Node 
+    {
+    #region ClassEnums
+    public enum AdResponse { Closed, Opened, Loaded, Started, Finished, Error, Rewarded }
+    #endregion
+    
+    #region Signals
+    [Signal] public delegate void AdCallback(AdResponse response);
```
Comment (ricardoalcantara) [inline admob-lib/AdMob.cs:12]: This Signal is not from the original GDScript version, why do you need that? Also is not in Godot Style. 
**Signals
Use the past tense to name signals**
https://docs.godotengine.org/en/stable/getting_started/scripting/gdscript/gdscript_styleguide.html#signals

File: admob-lib/AdMob.cs [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+    [Signal] public delegate void RewardedVideoLeftApplication();
+    [Signal] public delegate void RewardedVideoFailedToLoad(int error_code);
+    [Signal] public delegate void RewardedVideoOpened();
+    [Signal] public delegate void RewardedVideoStarted();
+    #endregion
+
+    #region Properties
+    [Export] bool IsReal = false;
+    [Export] bool BannerOnTop = true;
+    [Export] String BannerId = "ca-app-pub-3940256099942544/6300978111";
```
Comment (ricardoalcantara) [inline admob-lib/AdMob.cs:30]: The original file doesn't have the Ids hardcoded

File: admob-lib/AdMob.cs (hunk 3 above)
Comment (lefavilla) [reply admob-lib/AdMob.cs:4]: Oh ups. Not seen. Thats from an addon within VS will remove it if shin takes it.

**Comment by Shin-NiL** [comment]
Sorry for the delay guys.

I can't see the real benefit of adding this file to the library. As I said before, the AdMob.gd file only exists to help adding the node on the editor. 

The .cs script will add a complexity to the project that I, as main maintainer, can't deal with. I appreciate everyone's efforts, but I will have to reject the commit, sorry ;(

**Comment by lefavilla** [comment]
Okay. That's fine 👍 Usecase is very specific here.
Thanks for the cowork @Shin-NiL and @ricardoalcantara ! You helped me alot and I could benefit from your best practices.

Stay healthy!

**Pull request closed without merging**